golem-worker-executor-base/src/services/shard.rs

```diff
@@ -120,25 +171,47 @@ impl ShardServiceMock {
 
 #[cfg(any(feature = "mocks", test))]
 impl ShardService for ShardServiceMock {
-    fn assign_shards(&self, shard_ids: &HashSet<ShardId>) {
-        tracing::info!("ShardServiceMock::assign_shards {:?}", shard_ids)
+    fn is_ready(&self) -> bool {
+        tracing::info!("ShardServiceMock::is_ready");
+        true
+    }
+
+    fn assign_shards(&self, shard_ids: &HashSet<ShardId>) -> Result<(), GolemError> {
+        tracing::info!("ShardServiceMock::assign_shards {:?}", shard_ids);
+        Ok(())
     }
+
     fn check_worker(&self, worker_id: &WorkerId) -> Result<(), GolemError> {
         tracing::info!("ShardServiceMock::check_worker {:?}", worker_id);
         Ok(())
     }
+
     fn register(&self, number_of_shards: usize, shard_ids: &HashSet<ShardId>) {
         tracing::info!(
             "ShardServiceMock::register {} {:?}",
             number_of_shards,
             shard_ids
         )
     }
-    fn revoke_shards(&self, shard_ids: &HashSet<ShardId>) {
-        tracing::info!("ShardServiceMock::revoke_shards {:?}", shard_ids)
+
+    fn revoke_shards(&self, shard_ids: &HashSet<ShardId>) -> Result<(), GolemError> {
+        tracing::info!("ShardServiceMock::revoke_shards {:?}", shard_ids);
+        Ok(())
+    }
+
+    fn current_assignment(&self) -> Result<ShardAssignment, GolemError> {
+        tracing::info!("ShardServiceMock::current_assignment");
+        Ok(ShardAssignment::default())
     }
 
-    fn current_assignment(&self) -> ShardAssignment {
-        ShardAssignment::default()
+    fn opt_current_assignment(&self) -> Option<ShardAssignment> {
+        tracing::info!("ShardServiceMock::opt_current_assignment");
+        None
+    }
+}
+
+fn sharding_not_ready_error() -> GolemError {
```

Comment:
So as the `WorkerService` below can fail with this error now, the PR as is is not good enough yet - because the worker service (the actual service) needs to retry invocations until the sharding gets properly assigned. 

So basically I think the system needs to behave the same way as currently with `GoelmError::InvalidShardId` when sharding is not ready. Otherwise we will end up bubbling this error up to the user. 

Unfortunately we cannot just emit `InvalidShardId` because that contains a `ShardId` field which we cannot compute without knowing the number of shards.

So I propose to:
- hold this until the serialization tests are done, so we feel safer
- introduce a new error case somehow
- and as part of this PR, also handle it in the worker service
- additionally, probably the sharding 'readiness' should be checked in the beginning of the gRPC handlers, not randomly failing somewhere in the middle of handling them.